### PR TITLE
Allow users to login using their email address

### PR DESCRIPTION
**Pull request by SegiNyn**
As discussed, this pr is to allow users to login with an email address linked to their account in place of their username, using the same credentials. So the email address will be checked and accepted for login if the password matches the one they already use to login with their username. 
It also prevents using an email address as username. This is to avoid using an email address linked to another account as your username.

### CODE REVIEW

File: indico/modules/auth/forms.py [hunk 1]
```diff
@@ -46,22 +48,39 @@ def _check_not_blacklisted(form, field):
...
+    username = StringField(_('Username'), [DataRequired(), _check_existing_username, _check_use_own_email_as_username],
+                           filters=[_tolower])
     password = PasswordField(_('Password'), [DataRequired(), SecurePassword('set-user-password',
                                                                             username_field='username')],
                              render_kw={'autocomplete': 'new-password'})
     confirm_password = PasswordField(_('Confirm password'), [DataRequired(), ConfirmPassword('password')],
                                      render_kw={'autocomplete': 'new-password'})
 
+    def __init__(self, *args, **kwargs):
+        self.user = kwargs.pop('user', None)
```
Comment (ThiefMaster) [inline indico/modules/auth/forms.py:78]:
```suggestion
    def __init__(self, *args, user=None, **kwargs):
        self.user = user
```

the kwargs pop is a leftover from python 2 times where you could not mix specific keyword(-only) args and `**kwargs`

File: indico/modules/auth/providers.py [hunk 2]
```diff
@@ -30,6 +33,34 @@ def check_password(self, identity, password):
             session['insecure_password_error'] = error
         return True
 
+    def process_local_login(self, data):
+        identity = self.identity_model.query.filter(type(self).provider_column == self.name,
```
Comment (ThiefMaster) [inline indico/modules/auth/providers.py:37]: Remove the inheritance from `SQLAlchemyAuthProviderBase` and inherit from `AuthProviderBase` instead, and use the column/model names directly in here  instead of defining them on the class level (this was only useful when using the base class from flask-multipass).

Also, you can most likely write a single query that gets the local identities based on both the usual identifier match and email matches. That way you can avoid a large mount of code below.

Comment (SegiNyn) [reply indico/modules/auth/providers.py:37]: Cannot find `AuthProviderBase` is it `AuthProvider`?

If I use a query such as below
```
identity = (Identity.query.filter(Identity.provider == self.name,
                                          db.or_(Identity.identifier == data['identifier'],
                                                 UserEmail.email == data['identifier']),
                                          ~User.is_pending,
                                          ~UserEmail.is_user_deleted)
                                  .join(User)
                                  .join(UserEmail).first())
```
 Then I'll still need to do another query below to show that the user is logging in with an email address that isn't the primary email. Or we eliminate that part?

Comment (ThiefMaster) [reply indico/modules/auth/providers.py:37]: I'd check `data['identifier'] in user.secondary_emails` for that

Comment (SegiNyn) [reply indico/modules/auth/providers.py]: Okay

File: indico/modules/auth/forms.py [hunk 3]
```diff
@@ -46,22 +47,33 @@ def _check_not_blacklisted(form, field):
             raise ValidationError(error_msg)
 
 
+def _check_use_own_email_as_username(form, field):
+    if validate_email(field.data, check_dns=False):
+        if field.data not in form.user.all_emails:
+            raise ValidationError(_('Only emails addresses linked to your account can be used as a username.'))
+
+
```
Comment (SegiNyn) [inline indico/modules/auth/forms.py:55]: @ThiefMaster, @OmeGak, About this validation `_check_use_own_email_as_username` what we actually discussed was to prevent users from having characters such as '@' in their username and to do the validation on new usernames. 

But IMO this check would not be sufficient because it would prevent usernames like 'John@' which seems like a reasonable username. So what I did was to prevent users from using an email as an identifier that isn't their own email. @OmeGak raised a point yesterday that if a user uses an email as identifier then they can later remove that email from their emails and would then have an identifier that isn't their own email. 

If either of you have a strong opinion about definitely not allowing emails as usernames then I can remove this validation and just ensure no email is used as username. (For existing usernames to remain valid there'll be extra checks for if old_data != new_data)

Comment (ThiefMaster) [reply indico/modules/auth/forms.py:55]: I don't think `John@` needs to be allowed as a username.

TBH I'd even be OK with just allowing alphanumeric usernames at some point, but there's no need for this right now...

Comment (ThiefMaster) [reply indico/modules/auth/forms.py:55]: Should we just disallow emails as the username in general, and make the username an optional field? That way logging in with email/password would always work, and if a user wants to set a username they can...

Comment (SegiNyn) [reply indico/modules/auth/forms.py:55]: It should be ok not allowing emails as usernames. But isn't there a constraints on the identifier for it to always have a value? If so then they can continue to have usernames but login with email address if they want

Comment (ThiefMaster) [reply indico/modules/auth/forms.py:55]: Yes, it would need a change to the DB model to allow the Identity to have a null identifier when it's using the indico provider.

File: indico/modules/auth/providers.py [hunk 4]
```diff
@@ -30,6 +31,25 @@ def check_password(self, identity, password):
...
         return True
 
+    def process_local_login(self, data):
+        identity = (Identity.query.filter(Identity.provider == self.name,
+                                          db.or_(Identity.identifier == data['identifier'],
+                                                 UserEmail.email == data['identifier']),
+                                          ~User.is_pending,
+                                          ~UserEmail.is_user_deleted)
+                                  .join(User)
+                                  .join(UserEmail).first())
```
Comment (ThiefMaster) [inline indico/modules/auth/providers.py:41]: I don't think this is always correct. You query a single identity but I think we should consider edge cases where you have an existing username that looks like an email address.

For example, on my test instance I have an account with `a.m@cern.ch` and then created a local account with a username of `a.m@cern.ch`.

Now this query just gives me one identity, which may or may not be the correct one. That's why we need to use `.all()` here and iterate over the matched identities:

- If there are no identities --> NoSuchUser
- If the password check fails for all of them --> InvalidCredentials
- otherwise continue with the success logic for the first identity where the password was correct (I think we can safely ignore the case where the password is also correct for multiple identities, in that case it's almost certainly the same person just doing something dumb w/ multiple accounts and they'll probably open a support case eventually leading to merging accounts)

So basically this:

```py
    def process_local_login(self, data):
        # Query identities either by (user-settable) username or by email address.
        # In some very rare edge cases we can get more than one, e.g. when a user has an old identity
        # with a username that happens to be the email address of another user (this is no longer possible,
        # but existing accounts could have it).
        identities = (
            Identity.query
            .join(User)
            .join(UserEmail)
            .filter(Identity.provider == self.name,
                    db.or_(Identity.identifier == data['identifier'],
                           UserEmail.email == data['identifier']),
                    ~User.is_pending,
                    ~UserEmail.is_user_deleted)
            .order_by(Identity.id)
            .all()
        )
        # If we have no matching identities we fail with invalid username
        # XXX This is intentional, having an account on an Indico instance is generally not considered
        # secret information, and we want to give users the benefit of more verbose error messages.
        if not identities:
            raise NoSuchUser(provider=self)
        # From all the matching identities (usually just one), get one where the password matches, or
        # fail with invalid-password if there is none.
        if not (identity := next((ide for ide in identities if self.check_password(ide, data['password'])), None)):
            raise InvalidCredentials(provider=self)
        if data['identifier'] != identity.identifier and data['identifier'] in identity.user.secondary_emails:
            flash(_('You are logging in with a secondary email address. Please note that account related '
                    'notifications will always be sent to your primary email address. Go to "My profile" to '
                    'manage your emails.'), 'warning')
        auth_info = AuthInfo(self, identity=identity)
        return self.multipass.handle_auth_success(auth_info)
```

Feel free to copy this.

File: indico/modules/auth/providers.py [hunk 5]
```diff
@@ -5,20 +5,21 @@
...
 from indico.modules.auth import Identity, logger
 from indico.modules.auth.forms import LocalLoginForm
 from indico.modules.users import User
+from indico.modules.users.models.emails import UserEmail
+from indico.util.i18n import _
 from indico.util.passwords import validate_secure_password
 
 
-class IndicoAuthProvider(SQLAlchemyAuthProviderBase):
+class IndicoAuthProvider(AuthProvider):
```
Comment (ThiefMaster) [inline indico/modules/auth/providers.py:21]: I think we also need a unit test for the new logic considering this is rather sensitive logic.

This could either be done by using the test client and simulating actual requests, or simply by instantiating the `IndicoAuthProvider`, mocking out its `multipass` attribute, and then calling its `process_local_login` method and asserting that the mocked `self.multipass.handle_auth_success(auth_info)` has been called (in case of a successful login only)

Comment (SegiNyn) [reply indico/modules/auth/providers.py:21]: Okay, will add tests for this

File: indico/modules/auth/providers.py (hunk 4 above)
Comment (SegiNyn) [reply indico/modules/auth/providers.py]: Done

File: indico/modules/auth/providers.py (hunk 5 above)
Comment (SegiNyn) [reply indico/modules/auth/providers.py:21]: Done

**Comment by SegiNyn** [comment]
Hi @ThiefMaster, we'd like to add some template hooks so we can stop overriding some full files and it's related to changes made in this pr. Is it ok to include the hooks here rather than in a separate pr?

**Comment by ThiefMaster** [comment]
Fine for me, if they are too noisy I'll let you know and you can move them to a separate PR.

File: indico/modules/auth/templates/accounts.html [hunk 6]
```diff
@@ -32,6 +32,7 @@
                         <div id="credentials-form" {%- if not form.is_submitted() %} style="display: none;"{% endif %}>
                             {{ form_header(form, orientation='vertical', classes='no-block-padding') }}
                             {{ form_rows(form) }}
+                            {{ template_hook('below-account-form', include_form_style=false) }}
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/accounts.html:35]: I would pass the `user` here

File: indico/modules/auth/templates/accounts.html [hunk 7]
```diff
@@ -43,6 +44,7 @@
                     {% else %}
                         {{ form_header(form, orientation='vertical', classes='no-block-padding') }}
                         {{ form_rows(form) }}
+                        {{ template_hook('below-account-form', new_prefix=true, include_form_style=false) }}
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/accounts.html:47]: I'm not sure what `new_prefix` is supposed to mean, but it does not seem necessary since the code handling the template hook can simply check for `if user.local_identity`

File: indico/modules/auth/templates/reset_password.html [hunk 8]
```diff
@@ -21,6 +21,7 @@
 
         {{ form_header(form) }}
         {{ form_rows(form, widget_attrs=widget_attrs) }}
+        {{ template_hook('below-account-form') }}
```
Comment (ThiefMaster) [inline indico/modules/auth/templates/reset_password.html:24]: You use this in many different places, with no context whatsoever. I don't think this is a ghood idea. Either pass an argument indicating where it is, but maybe separate hooks (`reset-password-below-account-form` etc.) may be better.

**Review by ThiefMaster — COMMENTED** [review]
Leaving these comments since I already wrote them, but since I'll most likely add some additional things to this PR before merging it, I'm removing the commit adding them from here - please open a new PR to add those hooks.

**Review by ThiefMaster — APPROVED** [review]
